src/com/amazon/ion/benchmark/IonMeasurableWriteTask.java

```diff
@@ -195,4 +208,35 @@ public void closeWriter(IonWriter writer) throws IOException {
         writer.close();
     }
 
+    @Override
+    void generateWriteInstructionsElement(Consumer<WriteInstruction<IonWriter>> instructionsSink) throws IOException {
+        Iterable<AnyElement> elements;
+        if (options.limit == Integer.MAX_VALUE) {
+            try (IonReader reader = IonUtilities.newReaderBuilderForInput(options).build(options.newInputStream(inputFile))) {
+                elements = elementLoader.loadAllElements(reader);
+            }
+        } else {
+            List<AnyElement> limitedElements = new ArrayList<>();
+            try (IonReader reader = IonUtilities.newReaderBuilderForInput(options).build(options.newInputStream(inputFile))) {
+                Iterable<AnyElement> allElements = elementLoader.loadAllElements(reader);
+                int count = 0;
+                for (AnyElement element : allElements) {
+                    limitedElements.add(element);
+                    if (++count >= options.limit) break;
+                }
```

Comment:
So, what you're doing here is loading all elements, and then copying them one-by-one into a shorter list. One of the benefits of setting a limit is that if the data stream is too large, you can still load a subset of the data stream. However, this implementation will break if there is a data stream that is too large for the backing implementation. Instead, you can do something like this to ensure that we don't load more than the requested number of items from the data stream.

```suggestion
                int count = 0;
                while (count++ < options.limit && reader.next() != null) {
                    limitedElements.add(elementLoader.loadCurrentElement(reader));
                }
```

Reply:
Thanks for catching this.